### PR TITLE
Implementation of DCV Image Generation Module and EKS DCV Module

### PR DESCRIPTION
**Pull request by sauronalexander**
*Description of changes:*
This PR includes two modules:
- DCV image module: this module will create an ECR and build the DCV container image. In the DCV container, it will first grab the credentials from AWS Secrets Manager. It will create a virtual session. After the session is created, it will update the display number as well as the socket path on the node to ConfigMap where customers could checkout.
- DCV EKS module: this module will create necessary roles and k8s resources. It will create a new namespace in which all the roles, service accounts, config maps, role bindings and services related to dcv will be created. It will also create a DaemonSet in that namespace which contains the DCV pod template. Further more, it will add necessary node port to allowlist in the node security group. On module destruction, all the additional k8s resources will be deleted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by sauronalexander** [comment]
Update 12/9: Adding unit tests and README for dcv-image and dcv-eks module

**Comment by a13zen** [comment]
Does this support deployment into private subnets?

**Comment by sauronalexander** [comment]
> Does this support deployment into private subnets?

No, this is not included in this PR. I think it would be a separate module that creates a client VPN.

File: manifests/mlops/dcv-eks.yaml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+name: dcv-eks
+path: modules/visualization/dcv-eks
+parameters:
+  - name: DcvNamespace
```
[the remark this answers is not in the document]
Comment (sauronalexander) [reply manifests/mlops/dcv-eks.yaml:4]: Ack. Updating.

File: manifests/mlops/dcv-image.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+name: dcv-image
+path: modules/visualization/dcv-image
+parameters:
+  - name: DcvEcrRepositoryName
```
[the remark this answers is not in the document]
Comment (sauronalexander) [reply manifests/mlops/dcv-image.yaml:4]: Ack. Will update.

File: manifests/mlops/deployment.yaml [hunk 3]
```diff
@@ -7,6 +7,10 @@ groups:
   #   path: manifests/mlops/replicator-modules.yaml
   - name: core
     path: manifests/mlops/core-modules.yaml
+#  - name: dcv-image
```
[the remark this answers is not in the document]
Comment (sauronalexander) [reply manifests/mlops/deployment.yaml:10]: Yes, will uncomment

File: modules/visualization/dcv-eks/README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
+
+# DCV-EKS Module
+
+## Description
+
+This module creates necessary resources for DCV server to work.
+- Updates EKS node role to include permissions to access DCV license in S3
+- Updates security groups of the nodes to allow ingress traffic (TCP and UDP) to specific DCV nodeport
+- Creates a role for DCV pods to access AWSSecretsManager and kubernetes resources
+- Creates necessay k8s resources:
```
[the remark this answers is not in the document]
Comment (sauronalexander) [reply modules/visualization/dcv-eks/README.md:10]: Updated and resolving this comment.

File: modules/visualization/dcv-eks/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+#### Using AWS SecretsManager
+The key `dcv-cred-user` and `dcv-cred-passwd` need to exist in AWS SecretsManager.
+
+## Inputs/Outputs
+
+### Input Paramenters
+
+#### Required
+
+- `DcvImageUri` - DCV container image uri
```
[the remark this answers is not in the document]
Comment (sauronalexander) [reply modules/visualization/dcv-eks/README.md:31]: Ack. Updating.

File: modules/visualization/dcv-image/README.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
+# DCV-IMAGE Module
```
[the remark this answers is not in the document]
Comment (sauronalexander) [reply modules/visualization/dcv-image/README.md:1]: Added a couple of descriptions of this module.

File: modules/visualization/dcv-eks/README.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
+
+# DCV-EKS Module
```
[the remark this answers is not in the document]
Comment (sauronalexander) [reply modules/visualization/dcv-eks/README.md:2]: Added a couple of descriptions of this module.

File: manifests/mlops/optional-modules.yaml [hunk 8]
```diff
@@ -3,6 +3,9 @@ path: git::https://github.com/awslabs/idf-modules.git//modules/network/basic-cdk
 parameters:
   - name: internet-accessible
     value: true
+---
+name: addf-ecr
+path: modules/optionals/ecr/
```
[the remark this answers is not in the document]
Comment (sauronalexander) [reply manifests/mlops/optional-modules.yaml:8]: Will update to the reference in idf module.

File: modules/visualization/dcv-eks/README.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+information (display number and local mounting path) to configmap and aws 
+ssm parameter store. There will also be failure handling in DCV image. DCV 
+container will be restarted if the health check fails. The DCV session 
+configurations are outlined in dcv-image module.
+
+For user application pods to stream, user need to have two required inputs 
+to their pods:
+1. `DISPLAY` environment variable: usually the value is `:0`, but it is not 
+   guaranteed. User could check the actual display value in K8S ConfigMap 
+   `dcv::dcv-agent-config-map`.
```
Comment (srinivasreddych) [inline modules/visualization/dcv-eks/README.md:19]: Can you add the relevant kubectl command to check the value?

File: modules/visualization/dcv-eks/app.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+CfnOutput(
+    scope=dcv_eks_stack,
+    id="metadata",
+    value=dcv_eks_stack.to_json_string(
+        {
+            "DcvEksRoleArn": dcv_eks_stack.eks_admin_role.role_arn,
+            "DcvNamespace": dcv_namespace,
+            "DcvNodeport": dcv_node_port,
+            "DcvDisplayParameterName": ADDF_SSM_PARAMETER_STORE_DISPLAY_NAME,
```
Comment (srinivasreddych) [inline modules/visualization/dcv-eks/app.py:70]: Is it a good idea to have `ADDF` prefixed for SSM param? Instead, i would suggest the combination of `project_name`+`deployment_name`+`module_name` or `deployment_name`+`module_name` , which would let multiple instances of the same module deployment in an account/region... happy to discuss...

File: modules/visualization/dcv-eks/app.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+            "DcvSharedDirectoryParameterName": ADDF_SSM_PARAMETER_STORE_MOUNT_PATH_NAME,
```
Comment (srinivasreddych) [inline modules/visualization/dcv-eks/app.py:71]: Is it a good idea to have `ADDF` prefixed for SSM param? Instead, i would suggest the combination of `project_name`+`deployment_name`+`module_name` or `deployment_name`+`module_name` , which would let multiple instances of the same module deployment in an account/region... happy to discuss...

File: modules/visualization/dcv-eks/dcv-config/dcv-agent.yaml [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    spec:
+      restartPolicy: Always
+      serviceAccountName: dcv-agent-src-account
+      containers:
+        - name: nice-dcv-container
+          image: $IMAGE
+          command: [ "/usr/local/bin/run_script.sh" ]
+          imagePullPolicy: "Always"
+          env:
+            - name: ADDF_CONFIG_MAP_NAME
```
Comment (srinivasreddych) [inline modules/visualization/dcv-eks/dcv-config/dcv-agent.yaml:23]: Can you remove the ADDF prefix? Is it being referenced elsewhere? Try to use deployment_name and module_name wherever possible

File: modules/visualization/dcv-eks/modulestack.template [hunk 13]
```diff
@@ -0,0 +1,26 @@
+AWSTemplateFormatVersion: 2010-09-09
```
Comment (srinivasreddych) [inline modules/visualization/dcv-eks/modulestack.template:1]: we can remove this template if not needed

File: modules/visualization/dcv-image/modulestack.template [hunk 14]
```diff
@@ -0,0 +1,26 @@
+AWSTemplateFormatVersion: 2010-09-09
```
Comment (srinivasreddych) [inline modules/visualization/dcv-image/modulestack.template:1]: Can you remove this template if not needed?

File: modules/visualization/dcv-image/src/check_healthy.sh [hunk 15]
```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+
```
Comment (srinivasreddych) [inline modules/visualization/dcv-image/src/check_healthy.sh:2]: License file is missing here

File: modules/visualization/dcv-image/src/update_parameters.py [hunk 16]
```diff
@@ -0,0 +1,93 @@
+import json
```
Comment (srinivasreddych) [inline modules/visualization/dcv-image/src/update_parameters.py:1]: License file is missing

File: modules/visualization/dcv-image/src/update_parameters.py [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+ADDF_CONFIG_MAP_ENV_VAR_NAME = "ADDF_CONFIG_MAP_NAME"
+ADDF_APPLICATION_NAMES_ENV_VAR_NAME = "ADDF_APPLICATION_NAMES"
+ADDF_DCV_SESSION_INFO_DISPLAY_KEY = "x11-display"
+ADDF_DCV_DISPLAY_KEY_SUFFIX = "display"
+ADDF_CONFIG_MAP_PATCH_KEY = "data"
+ADDF_DCV_SESSION_NAME = "default-session"
+ADDF_SSM_PARAMETER_STORE_DISPLAY_NAME = "dcv-display"
+
+
+def get_display_number() -> str:
```
Comment (srinivasreddych) [inline modules/visualization/dcv-image/src/update_parameters.py:21]: Can you pls add docstring to the functions in this file?

File: modules/visualization/dcv-eks/README.md (hunk 9 above)
Comment (sauronalexander) [reply modules/visualization/dcv-eks/README.md:19]: Yes. Adding in next update.

File: modules/visualization/dcv-eks/app.py (hunk 10 above)
Comment (sauronalexander) [reply modules/visualization/dcv-eks/app.py:70]: Sure. I would follow your suggestion on the naming. Will reflect the naming in the README as well

File: modules/visualization/dcv-eks/dcv-config/dcv-agent.yaml (hunk 12 above)
Comment (sauronalexander) [reply modules/visualization/dcv-eks/dcv-config/dcv-agent.yaml:23]: This is more of internal usage. I would suggest to use a relatively shorter name for better readability. I would use prefix of `DCV_EKS` for some level of visibility here.

File: modules/visualization/dcv-eks/modulestack.template (hunk 13 above)
Comment (sauronalexander) [reply modules/visualization/dcv-eks/modulestack.template:1]: Correct. Sorry forgot to remove these two files. Removing for this and next comment.

File: modules/visualization/dcv-image/src/update_parameters.py (hunk 17 above)
Comment (sauronalexander) [reply modules/visualization/dcv-image/src/update_parameters.py:21]: Added.

File: modules/visualization/dcv-image/src/check_healthy.sh (hunk 15 above)
Comment (sauronalexander) [reply modules/visualization/dcv-image/src/check_healthy.sh:2]: Updated for this and next comment

File: modules/visualization/dcv-eks/app.py [hunk 18]
```diff
@@ -0,0 +1,76 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+from typing import cast
+
+from aws_cdk import App, CfnOutput, Environment
+
+from stack import ADDF_SSM_PARAMETER_STORE_DISPLAY_NAME, ADDF_SSM_PARAMETER_STORE_MOUNT_PATH_NAME, DcvEksStack
```
Comment (srinivasreddych) [inline modules/visualization/dcv-eks/app.py:9]: ADDF Prefix is still there

File: modules/visualization/dcv-eks/app.py (hunk 10 above)
Comment (srinivasreddych) [reply modules/visualization/dcv-eks/app.py:70]: ADDF Prefix is still there

**Review by srinivasreddych — DISMISSED** [review]

**Review by srinivasreddych — APPROVED** [review]

**Pull request merged**